File: backend/product/typescript/src/commons/validations/validation.ts
import { ZodType } from "zod";
import { ResponseException } from "../exceptions/response-exception";
import { ErrorMessage } from "../helpers/error-message";

export class Validation {
    static validate<T>(schema: ZodType, data: T): T {
        const result = schema.safeParse(data)
        if (result.error) {
            const errorMessages: ErrorMessage[] = []
            for (let i = 0; i < result.error.errors.length; i++) {
                const errorMessage: ErrorMessage = {
                    field: result.error.errors[i].path[0].toString(),
                    message: result.error.errors[i].message
                }
                errorMessages.push(errorMessage)
            }
            throw new ResponseException(400, errorMessages, "validation error")
        }

        return result.data
    }
}

File: backend/order/python/commons/validations/validation.py
from pydantic import BaseModel, ValidationError
from typing import Union, Type
from ..exception.response_exception import ResponseException
import json

def validate(schema: Type[BaseModel], data: dict) -> Union[BaseModel, ResponseException]:
    try:
        return schema.model_validate_json(json.dumps(data))
    except Exception as e:
        errorMessages = []
        for error in e.errors():
            field = str(error.get("loc"))
            field = field.replace("(", "").replace(")", "").replace("'", "").replace(",", ".").replace(" ", "")
            field = field[:-1] if field[-1] == "." else field
            errorMessage = {"field": field, "message": error.get("msg")}
            errorMessages.append(errorMessage)
        raise ResponseException(400, errorMessages, "validation error")
